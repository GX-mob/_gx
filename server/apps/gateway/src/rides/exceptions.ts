import {
  UnprocessableEntityException,
  NotFoundException,
  ForbiddenException,
} from "@nestjs/common";

export class RideNotFoundException extends NotFoundException {}
export class InvalidRideTypeException extends UnprocessableEntityException {}
export class UnsupportedAreaException extends UnprocessableEntityException {}
export class RideNoReadPermission extends ForbiddenException {}
