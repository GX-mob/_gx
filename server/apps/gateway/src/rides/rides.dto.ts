import {
  IRide,
  RideTypes,
  RidePayMethods,
  IRoute,
  IRoutePoint,
  IGetRideInfoDto,
  IGetRidePricesDto,
  ICreateRideDto,
  IPendencie,
  RideStatus,
} from "@shared/interfaces";
import {
  ValidateNested,
  IsNotEmpty,
  IsString,
  IsEnum,
  IsInt,
  IsArray,
} from "class-validator";
import { Exclude, Expose } from "class-transformer";
import { UserRoles } from "@shared/interfaces";

class Point implements IRoutePoint {
  @IsArray()
  coord!: [number, number];

  @IsNotEmpty()
  primary!: string;

  @IsNotEmpty()
  secondary!: string;

  @IsNotEmpty()
  district!: string;
}

class Route implements IRoute {
  @ValidateNested()
  start!: Point;

  @ValidateNested()
  waypoints?: Point[];

  @ValidateNested()
  end!: Point;

  @IsNotEmpty()
  path!: string;

  @IsInt()
  distance!: number;

  @IsInt()
  duration!: number;
}

export class GetRideInfoDto implements IGetRideInfoDto {
  @IsNotEmpty()
  pid!: IRide["pid"];
}

export class GetRidesPricesDto implements IGetRidePricesDto {
  @IsNotEmpty()
  area!: IRide["area"];

  @IsString()
  subArea?: IRide["subArea"];
}

export class CreateRideDto implements ICreateRideDto {
  @ValidateNested()
  route!: Route;

  @IsEnum(RideTypes)
  type!: RideTypes;

  @IsEnum(RidePayMethods)
  payMethod!: RidePayMethods;

  @IsNotEmpty()
  country!: IRide["country"];

  @IsNotEmpty()
  area!: IRide["area"];

  @IsNotEmpty()
  subArea!: IRide["subArea"];
}

export class RideInfoDto implements IRide {
  _id!: any;
  pid!: string;
  @Expose({ groups: [UserRoles.DRIVER] })
  voyager!: any;
  @Expose({ groups: [UserRoles.DRIVER] })
  route!: IRoute;
  /**
   * * 1 = Normal
   * * 2 = VIG - Very important gx
   */
  @Expose({ groups: [UserRoles.DRIVER] })
  type!: RideTypes;
  /**
   * * 1 = Money
   * * 2 = Credit card
   */
  @Expose({ groups: [UserRoles.DRIVER] })
  payMethod!: RidePayMethods;
  /**
   * Ride costs
   */
  costs!: IRide["costs"];

  @Exclude()
  country!: string;
  @Exclude()
  area!: string;
  @Exclude()
  subArea!: string;
  @Exclude()
  status!: RideStatus;
  @Expose({ groups: [UserRoles.VOYAGER] })
  driver?: any;
  @Expose({ groups: [UserRoles.VOYAGER] })
  pendencies?: IPendencie[];

  constructor(ride: IRide) {
    Object.assign(this, ride);
  }
}
