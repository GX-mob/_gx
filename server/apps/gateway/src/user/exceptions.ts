import {
  NotFoundException,
  UnprocessableEntityException,
  ConflictException,
} from "@nestjs/common";
import { HTTP_EXCEPTIONS_MESSAGES } from "@shared/http-exceptions";

export class UserNotFoundException extends NotFoundException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.USER_NOT_FOUND);
  }
}

export class WrongPasswordException extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.WRONG_PASSWORD);
  }
}

export class UnchangedPasswordException extends ConflictException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.UNCHANGED_DATA);
  }
}

export class ContactVerificationFailedException extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.CONTACT_VERIFICATION_FAILED);
  }
}

export class ContactRegistredException extends ConflictException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.CONTACT_ALREADY_REGISTRED);
  }
}

export class InvalidCPFException extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.INVALID_CPF);
  }
}

export class CPFRegistredException extends ConflictException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.CPF_REGISTRED);
  }
}

export class TermsNotAcceptedException extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.TERMS_NOT_ACCEPTED);
  }
}

export class PasswordRequiredException extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.PASSWORD_REQUIRED);
  }
}

export class NotOwnContactException extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.NOT_OWN_CONTACT);
  }
}

export class InvalidContactException extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.INVALID_CONTACT);
  }
}

export class RemoveContactNotAllowed extends UnprocessableEntityException {
  constructor() {
    super(HTTP_EXCEPTIONS_MESSAGES.REMOVE_CONTACT_NOT_ALLOWED);
  }
}
