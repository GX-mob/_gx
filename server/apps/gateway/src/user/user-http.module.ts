import { Module } from "@nestjs/common";
import { UserAuthModule } from "./auth/auth.module";
import { UserRegisterModule } from "./register/register.module";
import { UserManagementModule } from "./management/management.module";

@Module({
  imports: [UserAuthModule, UserRegisterModule, UserManagementModule],
})
export class UserHttpModule {}
