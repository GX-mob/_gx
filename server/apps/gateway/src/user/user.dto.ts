import {
  IsBoolean,
  IsDateString,
  IsNotEmpty,
  IsString,
  ValidateIf,
} from "class-validator";
import {
  IAuthPasswordDto,
  IContactDto,
  IContactVerificationCheckDto,
  IUser,
  IUserRegisterDto,
  UserRoles,
} from "@shared/interfaces";
import { Exclude } from "class-transformer";

export class ContactDto implements IContactDto {
  @IsNotEmpty()
  @IsString()
  contact!: string;
}

export class ContactVerificationCheckDto
  extends ContactDto
  implements IContactVerificationCheckDto {
  @IsNotEmpty()
  @IsString()
  code!: string;
}

export class AuthPasswordDto extends ContactDto implements IAuthPasswordDto {
  @IsNotEmpty()
  @IsString()
  password!: string;
}

export class UserRegisterDto
  extends ContactVerificationCheckDto
  implements IUserRegisterDto {
  @IsNotEmpty()
  @IsString()
  firstName!: string;

  @IsNotEmpty()
  @IsString()
  lastName!: string;

  @IsNotEmpty()
  @IsString()
  cpf!: string;

  @IsNotEmpty()
  @IsDateString()
  birth!: string;

  @IsNotEmpty()
  @IsBoolean()
  terms!: boolean;

  @IsString()
  password?: string;
}

export class UserDto implements IUser {
  _id!: string;
  pid!: string;
  firstName!: string;
  lastName!: string;
  phones!: string[];
  emails!: string[];
  cpf!: string;
  birth!: Date;
  avatar!: string;
  roles!: UserRoles[];
  averageEvaluation!: number;

  @Exclude()
  password!: string;

  constructor(partial: Partial<IUser>) {
    Object.assign(this, partial);
  }
}

export class UpdateProfileDto {
  @ValidateIf((o) => !o.lastName)
  @IsNotEmpty()
  @IsString()
  firstName?: string;

  @ValidateIf((o) => !o.firstName)
  @IsNotEmpty()
  @IsString()
  lastName?: string;
}

export class UpdatePasswordDto {
  @IsNotEmpty()
  @IsString()
  current!: string;

  @IsNotEmpty()
  @IsString()
  intended!: string;
}

export class Enable2FADto extends ContactDto {}

export class Disable2FADto {
  @IsNotEmpty()
  @IsString()
  password!: string;
}

export class RemoveContactDto extends ContactDto {
  @IsNotEmpty()
  @IsString()
  password!: string;
}
