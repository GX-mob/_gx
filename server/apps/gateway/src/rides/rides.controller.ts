import { AuthGuard, DAccount, Driver } from "@app/auth";
import { Account, EAccountRoles } from "@core/domain/account";
import {
  Body,
  ClassSerializerInterceptor,
  Controller,
  Get,
  Param,
  Post,
  SerializeOptions,
  UseGuards,
  UseInterceptors,
} from "@nestjs/common";
import { RideNoReadPermission, RideNotFoundException } from "./exceptions";
import {
  CreateRideDto,
  GetRideInfoDto,
  GetRidesPricesDto,
  RideInfoDto,
} from "./rides.dto";
import { RidesService } from "./rides.service";

@Controller("rides/")
@UseGuards(AuthGuard)
export class RidesController {
  constructor(readonly rideService: RidesService) {}

  @Get("prices-status/:area/:subArea?")
  getPricesStatusHandler(@Param() { area, subArea }: GetRidesPricesDto) {
    return this.rideService.getPricesOfRidesType(area, subArea);
  }

  @Driver()
  @UseInterceptors(ClassSerializerInterceptor)
  @SerializeOptions({
    excludePrefixes: ["_"],
    groups: [EAccountRoles.Driver],
  })
  @Get(":pid")
  async getRideDataHandler(
    @DAccount() account: Account,
    @Param() { pid }: GetRideInfoDto,
  ) {
    const driverPid = account.getID();
    const ride = await this.rideService.getRideByPid(pid);

    if (!ride) throw new RideNotFoundException();
    if (!ride.driver || ride.driver.pid !== driverPid)
      throw new RideNoReadPermission();

    return new RideInfoDto(ride);
  }

  @Post()
  @UseInterceptors(ClassSerializerInterceptor)
  @SerializeOptions({
    excludePrefixes: ["_"],
    groups: [EAccountRoles.Voyager],
  })
  async createRideHandler(
    @DAccount() account: Account,
    @Body() body: CreateRideDto,
  ) {
    const ride = await this.rideService.create(account, body);
    return new RideInfoDto(ride);
  }
}
