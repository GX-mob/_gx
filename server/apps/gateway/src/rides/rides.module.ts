import { Module } from "@nestjs/common";
import { SessionModule } from "@app/session";
import { RidesService } from "./rides.service";
import { RidesController } from "./rides.controller";

@Module({
  imports: [SessionModule],
  providers: [RidesService],
  controllers: [RidesController],
})
export class RidesModule {}
