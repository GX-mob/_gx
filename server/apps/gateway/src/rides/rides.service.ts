import { Injectable } from "@nestjs/common";
import { TCreateRideDto } from "@core/interfaces";
import {
  RideRepository,
  RideAreaConfigurationRepository,
} from "@app/repositories";
import { User } from "@core/domain/user";
import {
  IRide,
  RideCreate,
} from "@core/domain/ride";
import {
  IRideAreaConfiguration,
  IRideTypeConfiguration,
  RideAreas
} from "@core/domain/ride-areas";

@Injectable()
export class RidesService {
  
  private rideAreas!: RideAreas;

  constructor(
    private rideRepository: RideRepository,
    private rideAreaConfigurationRepository: RideAreaConfigurationRepository,
  ) {
    this.init();
  }

  private async init() {
    /**
     * Get and store all area and subArea configuration for all rides types and prices in memory
     */
    const rideAreas = await this.rideAreaConfigurationRepository.model
      .find()
      .lean<IRideAreaConfiguration>();

    this.rideAreas = new RideAreas(rideAreas);
  }

  async getRideByPid(pid: IRide["pid"]) {
    return this.rideRepository.find({ pid });
  }

  async create(user: User, data: TCreateRideDto) {
    return this.rideRepository.create(new RideCreate(user, this.rideAreas, data));
  }

  /**
   * Returns the rides types and respective prices of the requested area
   * @param area
   * @param subArea
   * @returns {IRideTypeConfiguration} Price list, price for provided ride type or undefined
   */
  getPricesOfRidesType(
    area: string,
    subArea?: string,
  ): IRideTypeConfiguration[] {
    return this.rideAreas.getPricesForArea(area, subArea);
  }
}
