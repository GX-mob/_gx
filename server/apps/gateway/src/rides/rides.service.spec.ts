/**
 * @group unit/service
 * @group unit/gateway/service
 * @group unit/gateway/rides/service
 */
import { Test } from "@nestjs/testing";
import { LoggerModule } from "nestjs-pino";
import faker from "faker";
import { ConfigModule, ConfigService } from "@nestjs/config";
import { CacheModule, CacheService } from "@app/cache";
import {
  RepositoryModule,
  RideRepository,
  PendencieRepository,
  RideAreaConfigurationRepository,
} from "@app/repositories";
import { SessionModule, SessionService } from "@app/session";
import { parseISO } from "date-fns";
import { IRide, RideTypes, IRideAreaConfiguration } from "@shared/interfaces";
import { RidesService } from "./rides.service";
import {
  mockRide,
  mockAreaConfiguration,
  mockRideTypeConfiguration,
  mockPendencie,
} from "@testing/testing";
import {
  InvalidRideTypeException,
  UnsupportedAreaException,
} from "./exceptions";
import { CreateRideDto } from "./rides.dto";
import { UserService } from "../user/user.service";

describe("RideService", () => {
  let ridesService: RidesService;
  let rideRepository: RideRepository;
  const { route } = mockRide();
  const rideType1 = mockRideTypeConfiguration();
  const rideType2 = mockRideTypeConfiguration({
    type: 2,
    available: true,
    perKilometer: 1.6,
    perMinute: 0.5,
    kilometerMultipler: 0.3,
    minuteMultipler: 0.2,
    overBusinessTimeKmAdd: 0.6,
    overBusinessTimeMinuteAdd: 0.5,
  });
  const prices = [
    mockAreaConfiguration({
      general: [rideType1, rideType2],
    }),
  ];

  const pendencieRepositoryMock = {
    model: { find: jest.fn() },
  };

  const rideAreaConfigurationRepositoryMock = {
    model: {
      find: () => ({
        lean: async (): Promise<IRideAreaConfiguration[]> => [...prices],
      }),
    },
  };

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({ isGlobal: true }),
        LoggerModule.forRoot(),
        SessionModule,
        CacheModule,
        RepositoryModule,
      ],
      providers: [SessionService, RidesService],
    })
      .overrideProvider(ConfigService)
      .useValue({ get() {} })
      .overrideProvider(CacheService)
      .useValue({})
      .overrideProvider(PendencieRepository)
      .useValue(pendencieRepositoryMock)
      .overrideProvider(RideAreaConfigurationRepository)
      .useValue(rideAreaConfigurationRepositoryMock)
      .compile();

    ridesService = moduleRef.get<RidesService>(RidesService);
    rideRepository = moduleRef.get<RideRepository>(RideRepository);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should be defined", () => {
    expect(ridesService).toBeDefined();
  });

  it("should has rides types loaded", async () => {
    const prices = await rideAreaConfigurationRepositoryMock.model
      .find()
      .lean();
    const areas: { [area: string]: IRideAreaConfiguration } = {};

    prices.forEach((price) => {
      areas[price.area] = price;
    });

    expect(ridesService.areas).toStrictEqual(areas);
  });

  describe("getRideByPid", () => {
    it("should call rideRepository", async () => {
      const pid = faker.random.alphaNumeric(12);
      const rideRepositoryGet = jest
        .spyOn(rideRepository, "get")
        .mockResolvedValue(null);

      await ridesService.getRideByPid(pid);

      expect(rideRepositoryGet).toBeCalledWith({ pid });
    });
  });

  describe("getRideStatusPrice", () => {
    it("should throw UnsupportedAreaException", async () => {
      expect(() =>
        ridesService.getPricesOfRidesType("NON-EXISTENT"),
      ).toThrowError(new UnsupportedAreaException());
    });

    it("should return the price list of an area", () => {
      const { area } = prices[0];

      expect(ridesService.getPricesOfRidesType(area)).toStrictEqual(
        prices[0].general,
      );
    });

    it("should return the price list of an area", () => {
      const { area } = prices[0];
      const [subArea] = Object.keys(prices[0].subAreas);

      expect(ridesService.getPricesOfRidesType(area, subArea)).toStrictEqual(
        prices[0].subAreas[subArea],
      );
    });

    it("should return the fallback to area due to an undefined subArea", () => {
      const { area } = prices[0];

      expect(ridesService.getPricesOfRidesType(area, "not-have")).toStrictEqual(
        prices[0].general,
      );
    });
  });

  describe("isBusinessTime", () => {
    it("should return true", () => {
      const dateTime = parseISO("2020-06-13 12:00");
      const dateTime2 = parseISO("2020-06-13 20:00");

      expect(
        ridesService.isBusinessTime("America/Maceio", dateTime),
      ).toBeTruthy();
      expect(
        ridesService.isBusinessTime("America/Maceio", dateTime2),
      ).toBeTruthy();
    });

    it("should return false", () => {
      const dateTime = parseISO("2020-06-13 11:00");
      const dateTime2 = parseISO("2020-06-13 22:00");
      const sunday = parseISO("2020-06-14 12:00");

      expect(
        ridesService.isBusinessTime("America/Maceio", dateTime),
      ).toBeFalsy();
      expect(
        ridesService.isBusinessTime("America/Maceio", dateTime2),
      ).toBeFalsy();
      expect(ridesService.isBusinessTime("America/Maceio", sunday)).toBeFalsy();
    });
  });

  describe("getCostsOfRideType", () => {
    it("should throw InvalidRideTypeException", () => {
      expect(() =>
        ridesService.getCostsOfRideType(prices[0].general, 3),
      ).toThrowError(new InvalidRideTypeException());
    });

    it("should return the prices", () => {
      expect(
        ridesService.getCostsOfRideType(prices[0].general, RideTypes.Normal),
      ).toStrictEqual(
        prices[0].general.find((price) => price.type === RideTypes.Normal),
      );
    });
  });

  describe("create", () => {
    function makeCrateRideDto(ride: IRide) {
      const createRideDto = new CreateRideDto();
      createRideDto.route = ride.route;
      createRideDto.type = ride.type;
      createRideDto.payMethod = ride.payMethod;
      createRideDto.country = ride.country;
      createRideDto.area = ride.area;
      createRideDto.subArea = ride.subArea;

      return createRideDto;
    }

    it("should create with pendencie", async () => {
      const ride = mockRide();
      const createRideDto = makeCrateRideDto(ride);
      const rideRepositoryCreateSpy = jest.spyOn(rideRepository, "create");

      rideRepositoryCreateSpy.mockImplementation(
        async (input) => input as IRide,
      );

      pendencieRepositoryMock.model.find.mockImplementation(() => ({
        lean: jest.fn().mockResolvedValue([]),
      }));

      await ridesService.create(ride.voyager._id, createRideDto);

      const expectBaseCosts = ridesService.getRideCosts(ride);
      const expectTotalBaseCosts =
        expectBaseCosts.distance.total + expectBaseCosts.duration.total;
      const [[{ costs }]] = rideRepositoryCreateSpy.mock.calls;

      expect(costs.base).toBe(expectTotalBaseCosts);
      expect(costs.total).toBe(expectTotalBaseCosts);
    });

    it("should create without pendencies", async () => {
      const ride = mockRide();
      const pendencie = mockPendencie({ issuer: ride.voyager._id });
      const createRideDto = makeCrateRideDto(ride);
      const rideRepositoryCreateSpy = jest.spyOn(rideRepository, "create");

      rideRepositoryCreateSpy.mockImplementation(
        async (input) => input as IRide,
      );

      pendencieRepositoryMock.model.find.mockImplementation(() => ({
        lean: jest.fn().mockResolvedValue([pendencie]),
      }));
      await ridesService.create(ride.voyager._id, createRideDto);

      const expectBaseCosts = ridesService.getRideCosts(ride);
      const expectTotalBaseCosts =
        expectBaseCosts.distance.total + expectBaseCosts.duration.total;
      const [[{ costs, pendencies }]] = rideRepositoryCreateSpy.mock.calls;

      expect(pendencies).toStrictEqual([pendencie]);
      expect(costs.base).toBe(expectTotalBaseCosts);
      expect(costs.total).toBe(expectTotalBaseCosts + pendencie.amount);
    });
  });

  function priceCalculationTest(
    func: "distancePrice" | "durationPrice",
    value: number,
    over: number,
  ) {
    describe(func, () => {
      it("should calculate", () => {
        const costsType1InBusiness = ridesService[func](
          value,
          prices[0].general[0],
          true,
        );
        const costsType2InBusiness = ridesService[func](
          value,
          prices[0].general[1],
          true,
        );
        const costsType1OutBusiness = ridesService[func](
          value,
          prices[0].general[0],
          false,
        );
        const costsType2OutBusiness = ridesService[func](
          value,
          prices[0].general[1],
          false,
        );

        const costsType1OverOutBusiness = ridesService[func](
          value + over,
          prices[0].general[0],
          false,
        );
        const costsType2OverOutBusiness = ridesService[func](
          value + over,
          prices[0].general[1],
          false,
        );

        const costsType1OverInBusiness = ridesService[func](
          value + over,
          prices[0].general[0],
          true,
        );
        const costsType2OverInBusiness = ridesService[func](
          value + over,
          prices[0].general[1],
          true,
        );

        expect(costsType1InBusiness).toMatchSnapshot();
        expect(costsType2InBusiness).toMatchSnapshot();

        expect(costsType1OutBusiness).toMatchSnapshot();
        expect(costsType2OutBusiness).toMatchSnapshot();

        expect(
          costsType1InBusiness.total < costsType1OutBusiness.total,
        ).toBeTruthy();
        expect(
          costsType2InBusiness.total < costsType2OutBusiness.total,
        ).toBeTruthy();

        expect(
          costsType1OverOutBusiness.total > costsType1OutBusiness.total,
        ).toBeTruthy();
        expect(
          costsType2OverOutBusiness.total > costsType2OutBusiness.total,
        ).toBeTruthy();
        expect(costsType1OverOutBusiness.aditionalForLongRide > 0).toBeTruthy();
        expect(costsType2OverOutBusiness.aditionalForLongRide > 0).toBeTruthy();

        expect(
          costsType1OverInBusiness.total > costsType1InBusiness.total,
        ).toBeTruthy();
        expect(
          costsType2OverInBusiness.total > costsType2InBusiness.total,
        ).toBeTruthy();
        expect(costsType1OverInBusiness.aditionalForLongRide > 0).toBeTruthy();
        expect(costsType2OverInBusiness.aditionalForLongRide > 0).toBeTruthy();
      });
    });
  }

  priceCalculationTest("distancePrice", route.distance, 10);
  priceCalculationTest("durationPrice", route.duration, 40);

  describe("getRideCosts", () => {
    it("should calculate ", () => {
      const mockCreateRideDto: any = {};
      mockCreateRideDto.route = { ...route };
      mockCreateRideDto.area = "AL";
      mockCreateRideDto.subArea = "maceio";
      mockCreateRideDto.type = RideTypes.Normal;

      const ridePrice = ridesService.getRideCosts(mockCreateRideDto);
      const isBusinessTime = ridesService.isBusinessTime(prices[0].timezone);
      const distancePrice = ridesService.distancePrice(
        route.distance,
        prices[0].general[0],
        isBusinessTime,
      );
      const durationPrice = ridesService.durationPrice(
        route.duration,
        prices[0].general[0],
        isBusinessTime,
      );

      expect(distancePrice).toStrictEqual(ridePrice.distance);
      expect(durationPrice).toStrictEqual(ridePrice.duration);
    });
  });
});
