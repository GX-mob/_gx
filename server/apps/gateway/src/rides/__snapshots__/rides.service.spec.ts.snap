// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RideService distancePrice should calculate 1`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 11,
}
`;

exports[`RideService distancePrice should calculate 2`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 16,
}
`;

exports[`RideService distancePrice should calculate 3`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 4,
  "total": 15,
}
`;

exports[`RideService distancePrice should calculate 4`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 6,
  "total": 22,
}
`;

exports[`RideService durationPrice should calculate 1`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 3,
}
`;

exports[`RideService durationPrice should calculate 2`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 5,
}
`;

exports[`RideService durationPrice should calculate 3`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 1,
  "total": 6,
}
`;

exports[`RideService durationPrice should calculate 4`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 2,
  "total": 10,
}
`;
